src/main/java/slimeknights/tconstruct/smeltery/recipe/AbstractCastingRecipe.java

```diff
@@ -0,0 +1,202 @@
+    return null;
+  }
+
+  @Override
+  public IRecipeType<?> getType() {
+    return type;
+  }
+
+  public int getCoolingTime() {
+    return this.coolingTime;
+  }
+
+  public int getFluidAmount() {
+    return this.fluid.getAmount();
+  }
+
+  public FluidStack getFluidStack() {
+    return this.fluid;
+  }
+
+  public Fluid getFluid() {
+    return this.fluid.getFluid();
+  }
+
+  public Ingredient getCast() {
+    return cast;
+  }
+
+  /**
+   * Does this recipe consume/destroy the cast ingredient.
+   * @return  Whether to destroy cast input.
+   */
+  public boolean consumesCast() {
+    return this.consumed;
+  }
+
+  /**
+   * Does this recipe's output get put into the input slot
+   * @return  Whether to switch to input after casting
+   */
+  public boolean switchSlots() {
+    return this.switchSlots;
+  }
+
+  /**
+   * Keeps casting recipes out of the recipe book.
+   */
+  @Override
+  public boolean isDynamic() {
+    return true;
+  }
+
+  public static class Serializer<T extends AbstractCastingRecipe> extends ForgeRegistryEntry<IRecipeSerializer<?>>
+    implements IRecipeSerializer<AbstractCastingRecipe> {
+    private final Serializer.IFactory<T> factory;
+
+    public Serializer(Serializer.IFactory<T> factoryIn) {
+      this.factory = factoryIn;
+    }
+
+    @Override
+    public T read(ResourceLocation recipeId, JsonObject json) {
+      Ingredient cast = Ingredient.EMPTY;
+      String s = JSONUtils.getString(json, "group", "");
+      boolean consumed = false;
+      boolean switchSlots = JSONUtils.getBoolean(json, "switchslots", false);
+      if (json.has("cast")) {
+        JsonElement jsonelement = JSONUtils.getJsonObject(json, "cast");
+        cast = Ingredient.deserialize(jsonelement);
+        consumed = JSONUtils.getBoolean(json, "castconsumed", false);
+      }
+
+      if (!json.has("fluidstack"))
+        throw new JsonSyntaxException("Missing fluid input definition!");
+      JsonObject jsonFluid = JSONUtils.getJsonObject(json, "fluidstack");
+      Fluid fluid = ForgeRegistries.FLUIDS.getValue(new ResourceLocation(JSONUtils.getString(jsonFluid, "fluid")));
+      int amount = JSONUtils.getInt(jsonFluid, "amount");
+      FluidStack fluidStack = new FluidStack(fluid, amount);
+      ItemStack item = new ItemStack(JSONUtils.getItem(json, "result"));
+      int coolingtime;
+      if (!json.has("coolingtime")) {
+        int time = 24;
+        int temperature = fluidStack.getFluid().getAttributes().getTemperature() - 300;
+        coolingtime = time + (temperature * fluidStack.getAmount()) / 1600;
+      }
+      else {
+        coolingtime = JSONUtils.getInt(json, "coolingtime");
+      }
+      return this.factory.create(recipeId, s, cast, fluidStack, item, coolingtime, consumed, switchSlots);
+    }
+
+    @Nullable
+    @Override
+    public T read(ResourceLocation recipeId, PacketBuffer buffer) {
+      // TODO: Magic number go away
+      String s = buffer.readString(32767);
```

Comment:
Max size of a short I think. The number is not magic, its just big.

Also, don't name this `s`, name it `group`